### ISSUE TITLE
pass various Timeout values on the command line (ESPTOOL-340)

### ISSUE THREAD

**Issue by moefear85** [espressif/esptool#686]
I'm facing the following issue: Failed to get PID of a device on /dev/tnt0, using standard reset sequence.

I'm using a rts/dtr capable pty<->pty bridge (https://github.com/lcgamboa/tty0tty) between esptool and a usb-uart bridge. The PID error always comes up, whereas the Exception only occasionally occurs, and it generally succeeds in the same way as if esptool directly communicated with the usb-uart. Without the pty<->pty bridge, no PID error occurs at all.

I suspect esptool's timings for rts/dtr are too tight, as it always expects to be directly connected to the target board, hence doesn't tolerate lag/latency when re-/setting the chip. Is it possible to extend esptool so that various timeout values can be passed on the command line? Even the 3 second timeout during flashing is occasionally limiting when flashing over a transparent wifi-bridge, with or without the socket:// option.

Even better, albeit understandably more difficult to implement, would be retries on individual slip packets if no reply or wrong reply is recieved, with the retry count also specifiable on the command line

**Comment by moefear85** [comment]
I'm aware of the innate CDC capability of the ESP32S2. However, not all ESP chips have it. So a USB-UART converter is still in many cases needed. Otherwise, why would you still provide support for rfc2217 and raw sockets?

More importantly, there are many other reasons why various middleware can be needed between esptool and the target chip, such as splitting the transfer stream (so one can observe the communication without having to switch back and forth between esptool and various other tools accessing the same port).

As I am new to git, I don't yet know how to create a pull request. I have however modified esptool.py. Please visit https://github.com/moefear85/esptool for further details. My modification would however still benefit from configurable/lax timing constraints, which is why I opened this feature request. Nonetheless, I got it working even without timeout configurability.

**Comment by radimkarnis** [comment]
I think there might be some misunderstanding. You mention [ESP32-S2 USB-CDC](https://docs.espressif.com/projects/esp-idf/en/v4.3/esp32s2/api-guides/usb-console.html) while I am talking about [ESP32-S3 and C3 USB Serial/JTAG Controller](https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-guides/usb-serial-jtag-console.html) - these are two different things.

Nevertheless, esptool mainly supports flashing through USB-UART converter AND on top of that supports the two previously mentioned USB implementations. In order for it to choose the right mode, it has to check the USB PID first. If `Failed to get PID of a device..` happens, a classic USB-UART procedure runs. That is why I am trying to explain that this "warning" is irrelevant to your situation. 

I am glad you could solve your problems and a proper PR with a thorough description is welcome. But to be honest, at this moment I am having a hard time understanding the connection between adding a udp socket and the PID warning.

**Comment by radimkarnis** [comment]
I will close this issue now, feel free to reopen if you think we should continue the discussion.

**Comment by moefear85** [comment]
Hi again,

I read your explanation about innate UART/JTAG controller in the esp32c3/s3, and how it falls back if it fails to get the PID. But that is off-topic, as the question wasn't whether certain chips have this fallback capability or not, rather it was why it was failing to get the PID in the first place, and my assumption that it had to do with timing constraints, which is why as a solution I offered the modification. The udp socket is not intended to solve the PID problem, nor did I claim it did. The problem is still there, but it doesn't interfere with successful programming. Resolving the warning requires customizing the timing constraints. I took note that this feature is planned for the future. Nice to hear.

The udp socket is a separate/independent contribution on my part, to enable easy remote flashing. My knowledge of git is currently insufficient to properly PR. If this prohibits a successful integration into esptool, that's not a problem. I'm able to use it with great flexibility. Thanks to esptool for being well-structured and open, which eased the implementation. If anybody else needs the feature, they know where to find it. It's openly accessible to the public on my github page.
